### ISSUE TITLE
linen.scan/remat_scan create params with shape containing 0 when scan length=1

### ISSUE THREAD

**Issue by evanatyourservice** [google/flax#2760]
Hello!

I am not sure if this is done by necessity or just by chance, but nn.scan and remat_scan output params with leading dim shape = 0 when scan length = 1. I noticed but didn't care until I ran into problems later on as it breaks the convention of counting params using prod(shape), which is used fairly often in deep learning code including distributed shampoo.

for example:
```
nn.remat_scan(
    ResBlockV2,
    lengths=(1,),
)()(x)
```
can output shape `(0, 2, 3, 4, 4)`, so the workaround for now would be

```
if self.n_layers > 1:
    return nn.remat_scan(
        ResBlockV2,
        lengths=(self.n_layers,),
    )()(x)
else:
    return ResBlockV2()(x)
```

Is it possible for a single loop scan to output leading dim shape = 1 or would there be negative consequences that I don't know about? If It's possible I think it'd let code be concise without having to do the workaround above while minimizing surprises from a param shape containing a 0.

### System information
- Flax, jax, jaxlib versions: 0.6.2 flax, 0.3.25 jax/jaxlib
- Python version: 3.9

**Comment by cgarciae** [comment]
Hey @evanatyourservice, I am not able to reproduce it with this simple code:

```python
import flax.linen as nn
import jax.numpy as jnp
import jax 

MLP = nn.remat_scan(
    nn.Dense, variable_axes={'params': 0},
    split_rngs={'params': True}, lengths=(1,))

x = jnp.ones((3, 5))
module = MLP(5)
y, variables = module.init_with_output(jax.random.PRNGKey(0), x)

print(jax.tree_map(jnp.shape, variables))
print(y.shape)
```
```
FrozenDict({
    params: {
        bias: (1, 5),
        kernel: (1, 5, 5),
    },
})
(3, 5)
```
Can you try to post a reproducible examples?

**Comment by evanatyourservice** [comment]
Hm okay I'll have to take a second look at my code, I'm busy so I'll just close this for now but come back if something looks fishy after a second look. Thank you!
